x-pack/filebeat/input/gcs/job.go

```diff
@@ -121,57 +131,207 @@ func (j *job) Timestamp() time.Time {
+	isStreamGzipped := false
+	// check if stream is gziped or not
+	buf, err := reader.Peek(3)
+	if err != nil {
+		if errors.Is(err, io.EOF) {
+			err = nil
+		}
+		return reader, false, err
+	}
+
+	// gzip magic number (1f 8b) and the compression method (08 for DEFLATE).
+	isStreamGzipped = bytes.Equal(buf, []byte{0x1F, 0x8B, 0x08})
 
-	_, err = data.ReadFrom(reader)
+	if !isStreamGzipped {
+		return reader, false, nil
+	}
+
+	gzReader, err := gzip.NewReader(reader)
 	if err != nil {
-		return nil, fmt.Errorf("failed to read data from object reader with error: %w", err)
+		return reader, false, err
 	}
 
-	return data, err
+	return gzReader, true, nil
 }
 
-func (j *job) createEventFields(message string, data []mapstr.M) mapstr.M {
-	return mapstr.M{
-		"message": message, // original stringified data
-		"log": mapstr.M{
-			"file": mapstr.M{
-				"path": j.objectURI,
-			},
-		},
-		"gcs": mapstr.M{
-			"storage": mapstr.M{
-				"bucket": mapstr.M{
-					"name": j.src.BucketName,
+// evaluateJSON, uses a bufio.NewReader & reader.Peek to evaluate if the
+// data stream contains a json array as the root element or not, without
+// advancing the reader
+func evaluateJSON(reader *bufio.Reader) (io.Reader, bool, error) {
+	eof := false
+	for i := 0; ; i++ {
+		b, err := reader.Peek((i + 1) * 5)
+		if errors.Is(err, io.EOF) {
+			eof = true
+		}
+		startByte := i * 5
+		for j := 0; j < len(b[startByte:]); j++ {
+			char := b[startByte+j : startByte+j+1]
+			switch {
+			case bytes.Equal(char, []byte("[")):
+				return reader, true, nil
+			case bytes.Equal(char, []byte("{")):
+				return reader, false, nil
+			case unicode.IsSpace(bytes.Runes(char)[0]):
+				continue
+			default:
+				return nil, false, fmt.Errorf("unexpected error: JSON data is malformed")
+			}
+		}
+		if eof {
+			return nil, false, fmt.Errorf("unexpected error: JSON data is malformed")
+		}
+	}
+}
+
+func (j *job) createEvent(message []byte, data []mapstr.M, offset int64) beat.Event {
+	event := beat.Event{
+		Timestamp: time.Now(),
+		Fields: mapstr.M{
+			"message": string(message), // original stringified data
+			"log": mapstr.M{
+				"offset": offset,
+				"file": mapstr.M{
+					"path": j.objectURI,
 				},
-				"object": mapstr.M{
-					"name":         j.object.Name,
-					"content_type": j.object.ContentType,
-					"json_data":    data, // objectified data
+			},
+			"gcs": mapstr.M{
+				"storage": mapstr.M{
+					"bucket": mapstr.M{
+						"name": j.src.BucketName,
+					},
+					"object": mapstr.M{
+						"name":         j.object.Name,
+						"content_type": j.object.ContentType,
+						"json_data":    data, // objectified data, if parseJSON == true, else its empty array
+					},
 				},
 			},
-		},
-		"cloud": mapstr.M{
-			"provider": "goole cloud",
-		},
-		"event": mapstr.M{
-			"kind": "publish_data",
+			"cloud": mapstr.M{
+				"provider": "goole cloud",
+			},
+			"event": mapstr.M{
+				"kind": "publish_data",
+			},
```

Comment:
I'm wondering if we can't
```suggestion
			"cloud": struct{ Provider string `json:"provider"` } {
				Provider: "google cloud",
			},
			"event": struct{ Provider string `json:"kind"` } {
				Kind: "publish_data",
			},
```
to save map allocs. Leave to your judgement.

Also note spelling of Google either way.